aten/src/ATen/cuda/CUDAGenerator.cpp

```diff
@@ -1,46 +1,148 @@
-#include <ATen/Config.h>
-
 #include <ATen/CUDAGenerator.h>
-#include <ATen/Context.h>
-#include <THC/THCTensorRandom.h>
-#include <stdexcept>
+#include <c10/cuda/CUDAFunctions.h>
+
+namespace at {
 
-// There is only one CUDAGenerator instance. Calls to seed(), manualSeed(),
-// initialSeed(), and unsafeGetTH() refer to the THCGenerator on the current
-// device.
+namespace cuda { namespace detail {
 
-THCGenerator* THCRandom_getGenerator(THCState* state);
+// Ensures we only call cudaGetDeviceCount only once.
+static std::once_flag num_gpu_init_flag;
 
-namespace at {
+// Total number of gpus in the system.
+static int64_t num_gpus;
 
-CUDAGenerator::CUDAGenerator(Context * context_)
-  : Generator(Device(DeviceType::CUDA)), context(context_)
-{
+// Ensures default_gens_cuda is initialized once.
+static std::deque<std::once_flag> cuda_gens_init_flag;
+
+// Default, global CUDA generators, one per GPU.
+static std::vector<std::shared_ptr<CUDAGenerator>> default_gens_cuda;
+
+/* 
+* Populates the global variables related to CUDA generators
+* Warning: this function must only be called once!
+*/
+static void initCUDAGenVector(){
+  num_gpus = c10::cuda::device_count();
+  cuda_gens_init_flag.resize(num_gpus);
+  default_gens_cuda.resize(num_gpus);
 }
 
-CUDAGenerator::~CUDAGenerator() {
-  // no-op Generator state is global to the program
+/**
+ * PyTorch maintains a collection of default generators that get
+ * initialized once. The purpose of these default generators is to
+ * maintain a global running state of the pseudo random number generation,
+ * when a user does not explicitly mention any generator.
+ * getDefaultCUDAGenerator gets the default generator for a particular
+ * cuda device.
+ */
+CUDAGenerator* getDefaultCUDAGenerator(DeviceIndex device_index) {
+  std::call_once(num_gpu_init_flag, initCUDAGenVector);
+  DeviceIndex idx = device_index;
+  if (idx == -1) {
+    idx = c10::cuda::current_device();
+  }
+  TORCH_CHECK(idx >= 0 && idx < num_gpus);
```

Comment:
nit: this check should be in the else branch, since a user can't trigger out of bounds `current_device` (that means we, framework, fucked up) (yes, this is the nittiest of nits)